components/odh-notebook-controller/e2e/helper.go

```diff
@@ -0,0 +1,74 @@
+package e2e
+
+import (
+	"context"
+	"crypto/tls"
+	"log"
+	"net/http"
+
+	routev1 "github.com/openshift/api/route/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+func (tc *testContext) waitForControllerDeployment(name string, replicas int32) error {
+	err := wait.Poll(resourceRetryInterval, resourceCreationTimeout, func() (done bool, err error) {
+
+		controllerDeployment, err := tc.kubeClient.AppsV1().Deployments(tc.testNamespace).Get(context.TODO(), name, metav1.GetOptions{})
+
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return false, nil
+			}
+			log.Printf("Failed to get %s controller deployment", name)
+			return false, err
+
+		}
+
+		for _, condition := range controllerDeployment.Status.Conditions {
+			if condition.Type == appsv1.DeploymentAvailable {
+				if condition.Status == v1.ConditionTrue && controllerDeployment.Status.ReadyReplicas == replicas {
+					return true, nil
+				}
+			}
+		}
+
+		log.Printf("Error in %s deployment", name)
+		return false, nil
+
+	})
+	return err
+}
+
+func (tc *testContext) getNotebookRoute() (*routev1.Route, error) {
+	nbRouteLookupKey := types.NamespacedName{Name: testNotebookName, Namespace: tc.testNamespace}
```

Comment:
Do not use `testNotebookName` as a global variable. Get the notebook name and namespace from the notebook object using:

```
getNotebookRoute(meta metav1.ObjectMeta)
```

Reply:
Not addressed, notebook name should passed as an argument, not hardcoded inside the helper function:

```go
opts := []client.ListOption{
    client.InNamespace(tc.testNamespace),
    client.MatchingLabels{"notebook-name": "thoth-minimal-oauth-notebook"}}
```

Reply:
The reason to not pass the `meta` object to these functions is we would need to do a `Get` request for Notebook object for every test/sub-test.  Its also not possible to pass context between the test suites (create to delete)

I updated and restructured the code to take into consideration testing of multiple Notebooks. Introduced `NotebookContext` that maps to test Notebook, such that entire test suite is run for every notebook inside `[]NotebookContext` list. Anyone adding a test notebook just needs to add Notebook Spec to the context. Following is an example test run:
```
    --- PASS: TestE2ENotebookController/create (234.47s)
        --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook (231.47s)
            --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Creation_of_Notebook_instance (10.11s)
            --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Notebook_Route_Validation (10.04s)
            --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Notebook_Statefulset_Validation (10.04s)
            --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Notebook_OAuth_sidecar_Validation (0.04s)
            --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Verify_Notebook_Traffic (10.24s)
            --- PASS: TestE2ENotebookController/create/thoth-minimal-oauth-notebook/Verify_Notebook_Culling (191.00s)
```
As seen above every subtest is mapped to corresponding test notebook